tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ApalacheBuilder.scala

```diff
@@ -0,0 +1,57 @@
+package at.forsyte.apalache.tla.typecomp.subbuilder
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp.BuilderUtil._
+import at.forsyte.apalache.tla.typecomp._
+import at.forsyte.apalache.tla.typecomp.unsafe.UnsafeApalacheBuilder
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * Type-safe builder for ApalacheOper expressions.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+trait ApalacheBuilder extends UnsafeApalacheBuilder {
```

Comment:
Safe builders always extend unsafe builders, then you can do e.g.
```
binaryFromUnsafe(lhs, rhs)(_unsafeMethod)
```
This is why all unsafe methods are `protected`

Reply:
I don't understand why extending is required to do this. Couldn't you just import the methods of the unsafe builder so they are available in the safe one?

Reply:
I see you're already using this extension approach all over the builder, so of course not something to change here, but I don't understand the rationale for extending here vs. just importing.

Reply:
It's interface locking. Extension + `protected` makes it so that you can't use the unsafe methods.

Reply:
I also noticed this on a previous review but didn't raise it because it was already used in several places. Since Shon also raised this, I think it's worth discussing.

> It's interface locking. Extension + `protected` makes it so that you can't use the unsafe methods.

I don't follow. By "use", do you mean within the type-safe builder or outside?

Reply:
I think what we are discussing here is inheritance vs. composition. AFAIK, the common rule is to prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance#:~:text=To%20favor%20composition%20over%20inheritance,and%20creating%20a%20family%20tree.). Not sure, whether it is critical here.

Reply:
Yes, totally. I tried to explain why I think it's critical here: https://github.com/informalsystems/apalache/pull/1932#issuecomment-1177384054